rustls/src/suites.rs

```diff
@@ -216,6 +213,31 @@ pub enum ConnectionTrafficSecrets {
     },
 }
 
+impl ConnectionTrafficSecrets {
+    /// Convert three slices to three fixed sized arrays, panicking if any slice is not the correct
+    /// constant size length.
+    pub fn slices_to_arrays<const NK: usize, const NS: usize, const NI: usize>(
+        k: &[u8],
+        s: &[u8],
+        i: &[u8],
+    ) -> ([u8; NK], [u8; NS], [u8; NI]) {
+        (
+            Self::slice_to_array(k),
+            Self::slice_to_array(s),
+            Self::slice_to_array(i),
+        )
+    }
+
+    /// Convert a slice to a fixed sized array, panicking if the slice is not the correct
+    /// constant size length.
+    pub fn slice_to_array<const N: usize>(slice: &[u8]) -> [u8; N] {
+        // this is guaranteed true because `ConnectionTrafficSecrets` items and
+        // `key_block_shape()` are in agreement.
```

Comment:
I think this comment needs to be adjusted, and perhaps indicates that these helpers aren't as generic as they might seem? Open to suggestions for reworking

Reply:
Thoughts on doing this instead?

```rust
pub enum ConnectionTrafficSecrets {
  Aes128Gcm { key: AeadKey, iv: Iv }
  Aes256Gcm { key: AeadKey, iv: Iv }
  Chacha20Poly1305 { key: AeadKey, iv: Iv }
}
```

It will require a bit of churn in downstream https://github.com/hapsoc/ktls/blob/main/src/ffi.rs#L131

Reply:
> Thoughts on doing this ~instead~ later?

I'm going to run with this but think it might make sense to do separately from removing the `secret_extraction` feature.

Reply:
@ctz WDYT about this as a starting point: https://github.com/cpu/rustls/commit/4e09b6da8a460266de15a439729b43e980343b70 I think it's approaching what you had in mind, but there might be some ways to make it more ergonomic for crate external implementations.

Reply:
Let's land this version and I'll pull out the rework mentioned above as a separate PR to iterate on.

Reply:
https://github.com/rustls/rustls/pull/1497